### DISCUSSION TITLE
How to calculate maxFeePerGas/maxPriorityFeePerGas

### DISCUSSION THREAD

**Discussion by jackmellis** [wevm/viem#239] [Question] [answered]
We're working on migrating from ethers to viem (going good so far) but I can't figure this one out.

With ethers we could use `provider.getFeeData()` to get the `maxFeePerGas` and `maxPriorityFeePerGas` values, then use them to get the gas estimate. If we couldn't get these values, we'd assume we need to do a legacy transaction. Something like this:

```ts
const get1559GasFees = async ({ provider }: { provider: Provider }) => {
  try {
    return await provider.getFeeData();
  } catch {
    return {
      maxFeePerGas: null,
      maxPriorityFeePerGas: null,
    };
  }
};

let maxFeePerGas: BigNumber = null;
let maxPriorityFeePerGas: BigNumber = null;
let gasEstimate: BigNumber = null;

try {
  const fees = await get1559GasFees({
    provider: contract.provider as Provider,
  });
  maxFeePerGas = fees.maxFeePerGas;
  maxPriorityFeePerGas = fees.maxPriorityFeePerGas;
  // Estimate EIP-1559 gas
  gasEstimate = await contract.estimateGas[method](...args, {
    maxFeePerGas,
    maxPriorityFeePerGas,
  });
} catch {
  // EIP-1559 not supported/failed
  try {
    maxFeePerGas = null;
    maxPriorityFeePerGas = null;
    gasEstimate = await contract.estimateGas[method](...args);
  } catch {
    // Failed to estimate gas
  }

  return { gasEstimate, maxFeePerGas, maxPriorityFeePerGas };
```

Then we'd subsequently call the contract method with the resulting values:
```ts
await contract[method](...args, {
  maxFeePerGas, // may be null
  maxPriorityFeePerGas, // may be null
  gasLimit: bumpBy5Percent(gasEstimate), // the gas estimate for either a 1559 or a legacy txn
});
```

With Viem I can see that we can pass in `maxFeePerGas` and `maxPriorityFeePerGas` for an EIP-1559 transaction, or `gasLimit` for a legacy transaction. But it's not clear a) how I get those 1559 values, or b) whether I should still be passing gasLImit for a 1559 transaction?

```ts
const gasEstimate = await client.estimateContractGas({
  address,
  abi,
  functionName,
  args,
  account,
  maxFeePerGas, // ?
  maxPriorityFeePerGas, // ?
});

return bumpBy5Percent(gas) // is this still the gasLimit arg for a 1559 txn?
```

I'm really just hoping for some clarification on all this 😊

**Comment by jxom** [comment, marked as answer]
Are you using a JSON-RPC Account or a Local Account (private key, mnemonic, etc)?

For a JSON-RPC Account, it is recommended to defer fee calculation to the wallet (as they have better mechanisms to determine best fee data), meaning you don't need to pass through fee data at all. 
For a Local Account, we do fee calculations [automatically](https://github.com/wagmi-dev/viem/blob/main/src/utils/transaction/prepareRequest.ts#L50-L82) in the internal `prepareRequest` util – however, if you want to do something more custom, we can extract this out into a utility quite easily. Would you rather do this manually and we extract it out into a util?

**Reply by 0xsambugs** [reply]
Hey everyone! Just wanted to give my 2 cents. I'm currently migrating from Ethers to Viem, and the only thing I couldn't figure out yet is how to replace the `getFeeData`. That functions returns `maxFeePerGas` and `maxPriorityFeePerGas`, and I'd love to be able to have the same with Viem

A utility would be more than enough for me 💪

**Reply by quagliero** [reply]
A utility would be great. For example when wanting to sweep a wallet balance to zero I need to know the exact amounts to pass in so the tx won't fail (e.g. send wallet balance minus gas costs).

**Reply by melonges** [reply]
I have similar problem. I want to call  smart contract method with sending amount of the all available native balance. I got the error "An error occurred while executing: gas required exceeds allowance"

**Reply by kanishkatn** [reply]
I am having the same problem as well. +1 for the utility!

**Reply by gvash** [reply]
+1 for a utility function

**Reply by nikola-bozin-txfusion** [reply]
+1 for this feature!
